### DISCUSSION TITLE
How many connections when share a account in ThreadPoolExecutor?

### DISCUSSION THREAD

**Discussion by HeHongyeFY** [ecederstrand/exchangelib#1344] [Q&A]
I'm new in exchangelib, and I have some confusions about the account.protocol.

when I share the single account instance run in ThreadPoolExecutor, how many connections would established?
can one account.protocol's session be used in multiple threads at same time,  whether it be blocked by SESSION_POOL?
when I set CONNECTIONS_PER_SESSION=10,  How does it work?

I want clarify the mechanism！
Is there anyone explain it for me?
appreciate very much!!!

    # The maximum number of sessions (== TCP connections, see below) we will open to this service endpoint. Keep this
    # low unless you have an agreement with the Exchange admin on the receiving end to hammer the server and
    # rate-limiting policies have been disabled for the connecting user. Changing this setting only makes sense if
    # you are using threads to run multiple concurrent workers in this process.
    SESSION_POOLSIZE = 1
    # We want only 1 TCP connection per Session object. We may have lots of different credentials hitting the server and
    # each credential needs its own session (NTLM auth will only send credentials once and then secure the connection,
    # so a connection can only handle requests for one credential). Having multiple connections per Session could
    # quickly exhaust the maximum number of concurrent connections the Exchange server allows from one client.
    CONNECTIONS_PER_SESSION = 1

```python3
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import logging
import logging.config
from concurrent.futures import ThreadPoolExecutor
from exchangelib import Credentials, Configuration, Account, DELEGATE

def main():
    pool = ThreadPoolExecutor(max_workers=100)
    config = Configuration(
        credentials=Credentials(
            username='',
            password=''
        ),
        max_connections=10,
        server='192.190.40.50'
    )
    client = Account(
            primary_smtp_address='administrator@spinfo.exchange2016.com',
            access_type=DELEGATE,
            config=config
        )
    items = list()
    resp = client.inbox.all().values('id','changekey')
    for item in resp:
        items.append(item)
    for item in items:
        pool.submit(client.fetch, ids=[(item['id'], item['changekey'])], only_fields=['mime_content'])
        pool.shutdown()

if __name__ == "__main__":
    main()
```

**Comment by ecederstrand** [comment]
Protocol instances are cached and reused across multiple accounts. The caching is based on the server's EWS endpoint and the credentials used for authentication, so multiple accounts hosted on the same server will share a single protocol instance. This happens in https://github.com/ecederstrand/exchangelib/blob/7da64d79589b8a9fd34067d7ef53edc119f414b3/exchangelib/protocol.py#L397

By default, only one connection is opened per protocol instance. As the comment states, you can increase that limit on agreement with the server admins. Additionally, if you're using streaming subscriptions, then you'll need at least two connections, as described in https://ecederstrand.github.io/exchangelib/#synchronization-subscriptions-and-notifications

**Reply by HeHongyeFY** [reply]
Excuse me!

Do you mean I need increase the SESSION_POOLSIZE number?

I still have a doubt.

Look at my demo code, By default, when make requests,  protocol.get_session() execute,  another thread will block,  because there is only one session is occupied.
```python3
for item in items:
        pool.submit(client.fetch, ids=[(item['id'], item['changekey'])], only_fields=['mime_content'])
pool.shutdown()
```
requests.Session and it's connetionpool can be reused in multiple threads,  why not use it directly?
Why is SESSION_POOL necessary?

Is misunderstanding exist?
I want unserstand the design.

Thank you very much!

**Reply by ecederstrand** [reply]
If you want to run threaded, then you'll need to increase `SESSION_POOLSIZE` to the size of the thread pool to take advantage of using threads. Otherwise, all threads will have to compete for the same connection.

**Reply by HeHongyeFY** [reply]
respect！
